### PR TITLE
Speed up XContent Collection Parsing

### PR DESCRIPTION
**Pull request by original-brownbear**
1. Get rid of the capturing lambda on the hot path that inlines very badly
2. Remove as many bounds checks as possible, thereby reducing method size and improving inlining

### CODE REVIEW

**Review by pugnascotia — APPROVED** [review]
Looks reasonable, though I'm not very familiar with this part of the code 😅

**Review by jaymode — APPROVED** [review]
LGTM

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java [hunk 1]
```diff
@@ -263,117 +263,141 @@ public CharBuffer charBufferOrNull() throws IOException {
...
-        for (; token == XContentParser.Token.FIELD_NAME; token = parser.nextToken()) {
-            // Must point to field name
-            String fieldName = parser.currentName();
-            // And then the value...
-            parser.nextToken();
-            T value = mapValueParser.apply(parser);
-            map.put(fieldName, value);
-        }
-        return map;
+        return token == Token.FIELD_NAME;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java:348]: > The behavior was already that way, so no need for you to change it.

Yea this is+was really weird. I actually tried changing it here initially but we made use of this behaviour in a bunch of places (tests mostly) and it was not trivial to adjust those (would've blown up the PR size a lot) so I kept it for now.

**Comment by original-brownbear** [comment]
Thanks Rory+Jay!

**Pull request merged**